### PR TITLE
esp32s2: implement canio

### PR DESCRIPTION
**Pull request by jepler**
This is complete enough to build and not simply crash when creating a `canio.CAN` object.  No other testing performed as yet, so I'm entering this as a draft.  TODOs include:
 - [x] Test transmission
 - [x] Test reception in loopback mode
 - [x] Test reception filtering in loopback mode
 - [x] Test interoperation with other CAN implementations

### CODE REVIEW

**Comment by tannewt** [comment]
Please merge this into `main` rather than `6.0.x`. We'll release it in 6.1

**Comment by jepler** [comment]
In my testing, transmission works on "loopback", though an external wire from RX to TX is also needed.  There are problems around message reception and filtering.

**Comment by jepler** [comment]
Filters aren't working.  This may be a problem in esp-idf, so I've filed an issue there ^^

**Comment by jepler** [comment]
I've successfully worked around the apparent code optimization problem and tested that this can interoperate with the SAM E54 canio implementation.

File: ports/esp32s2/Makefile [hunk 1]
```diff
@@ -331,10 +331,10 @@ $(BUILD)/firmware.uf2: $(BUILD)/circuitpython-firmware.bin
 	$(Q)$(PYTHON3) $(TOP)/tools/uf2/utils/uf2conv.py -f 0xbfdd4eee -b 0x0000 -c -o $@ $^
 
 flash: $(BUILD)/firmware.bin
-	esptool.py --chip esp32s2 -p $(PORT) --no-stub -b 460800 --before=default_reset --after=hard_reset write_flash $(FLASH_FLAGS) 0x0000 $^
+	esptool.py --chip esp32s2 -p $(PORT) -b 460800 --before=default_reset --after=hard_reset write_flash $(FLASH_FLAGS) 0x0000 $^
```
Comment (tannewt) [inline ports/esp32s2/Makefile:334]: Please leave this. `--no-stub` is required for the native USB bootloader.

File: ports/esp32s2/common-hal/canio/CAN.c [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+#include "hal/twai_types.h"
+
+STATIC bool reserved_can;
+
+twai_timing_config_t get_t_config(int baudrate) {
+    switch(baudrate) {
+    case 1000000:
+    {
+        twai_timing_config_t t_config = TWAI_TIMING_CONFIG_1MBITS();
+        return t_config;
```
Comment (tannewt) [inline ports/esp32s2/common-hal/canio/CAN.c:47]: Why can't you:

```suggestion
        return TWAI_TIMING_CONFIG_1MBITS();
```

File: ports/esp32s2/common-hal/canio/CAN.c [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    }
+    if (silent) {
+        g_config.mode = TWAI_MODE_LISTEN_ONLY;
+    }
+    self->g_config = g_config;
+
+    {
+    twai_filter_config_t f_config = TWAI_FILTER_CONFIG_ACCEPT_ALL();
+    self->f_config = f_config;
+    }
```
Comment (tannewt) [inline ports/esp32s2/common-hal/canio/CAN.c:147]: Why this form?

File: ports/esp32s2/common-hal/canio/CAN.c [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+void common_hal_canio_can_check_for_deinit(canio_can_obj_t *self) {
+    if (common_hal_canio_can_deinited(self)) {
+        raise_deinited_error();
+    }
+}
+
+void common_hal_canio_can_deinit(canio_can_obj_t *self)
+{
+    if (self->tx_pin) {
+        (void)twai_stop();
```
Comment (tannewt) [inline ports/esp32s2/common-hal/canio/CAN.c:283]: Why `(void)`?

**Review by tannewt — CHANGES_REQUESTED** [review]
Nothing major. Just a couple spots that need explanation. Good job getting it going!

File: ports/esp32s2/common-hal/canio/CAN.c (hunk 2 above)
Comment (jepler) [reply ports/esp32s2/common-hal/canio/CAN.c:47]:
```
components/hal/include/hal/twai_types.h:#define TWAI_TIMING_CONFIG_1MBITS()     {.brp = 4, .tseg_1 = 15, .tseg_2 = 4, .sjw = 3, .triple_sampling = false}
```
This macro expands to a braced initializer.  In C, `return {.brp = 4, ...};` is not syntactically valid (nor can you assign it, which is why I wrote squirrely code below when assigning f_config as well).  Typical error if you try it:
```
common-hal/canio/CAN.c: In function 'get_t_config':
common-hal/canio/CAN.c:45:16: error: expected expression before '{' token
         return TWAI_TIMING_CONFIG_1MBITS();
                ^~~~~~~~~~~~~~~~~~~~~~~~~
```

File: ports/esp32s2/common-hal/canio/CAN.c (hunk 4 above)
Comment (jepler) [reply ports/esp32s2/common-hal/canio/CAN.c:283]: twai_stop returns an `esp_error_t`.  However, nothing useful can be done in the presence of an error.  For instance, we can't throw here.  While ignored return value diagnostics are apparently not enabled, casting the non-void return to void is a way to be explicit that we discard the return value.

File: ports/esp32s2/common-hal/canio/CAN.c (hunk 2 above)
Comment (tannewt) [reply ports/esp32s2/common-hal/canio/CAN.c:47]: Ah! Makes sense then. Please add a comment in the code detailing that.

File: ports/esp32s2/Makefile (hunk 1 above)
Comment (jepler) [reply ports/esp32s2/Makefile:334]: This should not have been added.  Will revert.  It's good to understand why --no-stub was specified, thanks for the explanation.

**Review by tannewt — APPROVED** [review]
Looks good! Thank you!

**Pull request merged**